keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1129 @@
+need to over-provision for each container’s peak usage. 
+
+
+### Notes/Constraints/Caveats
+
+
+1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
+2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
+turn it on by specifying the relevant resource section at pod-level in the Pod specification.
+
+
+### Risks and Mitigations
+
+Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
+Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
+as the feature matures and potential issues are addressed.
+
+
+## Design Details
+
+
+### Components/Features requiring changes
+
+
+#### PodSpec and PodStatus changes
+
+New field in `PodSpec` and `PodStatus`
+
+```
+type PodSpec struct {
+...
+  // Compute resource requirements.
+  // +optional
+  Resources ResourceRequirements
+}
+
+type PodStatus struct {
+...
+  // Compute resource requirements.
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+
+#### PodSpec Validation
+
+* **Scenario 1: Only container-level resources are specified (which is already supported without this KEP)**
+
+  ```yaml
+  apiVersion: v1
+  kind: Pod
+  ...
+  spec:
+    containers:
+    - name: container-1
+      resources:
+        requests:
+          memory: 100Mi
+          cpu: 1
+    - name: container-2
+  ...
+  ```
+
+  No changes required to support this.
+
+* **Scenario 2: Only pod-level resources are specified**
+
+  ```yaml
+  apiVersion: v1
+  kind: Pod
+  spec:
+    resources:
+      requests:
+        memory: 100Mi
+        cpu: 1
+  ...
+    containers:
+    - name: container-1
+    ...
+    - name: container-2
+  ...
+  ```
+
+  Following validation logic needs to be added for `resources` specified in the PodSpec:
+
+  1. Only CPU and memory resources are supported in pod level resource specifications, hence invalidate pod-level resource specification for other resources 
+  like gpu, storage, etc.
+
+* **Scenario 3: Both pod-level and container level resources are specified.**
+
+  ```yaml
+  apiVersion: v1
+  kind: Pod
+  ...
+  spec:
+  resources:
+    requests:
+      memory: 100Mi
+      cpu: 1
+    containers:
+    - name: container-1
+        resources:
+          requests:
+            memory: 100Mi
+            cpu: 1
+    ...
+    - name: container-2
+  ...
+  ```
+
+  Following validation logic needs to be added:
+
+  1. requests
+      * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu])` is valid.
```

Comment:
Does this mean "calculate the implicit pod request, in the standard way, and ensure that pod requests is at least as much" ?

Reply:
Yeah, that was my interpretation.

Reply:
I think this can section benefit from defining some terms for clarity, e.g., 
 - "Minimum effective pod request" for the implicit pod request from all the init/sidecar/app containers
 - `pod.spec.resources.requests[memory/cpu] >= minimum_effective_pod_request`